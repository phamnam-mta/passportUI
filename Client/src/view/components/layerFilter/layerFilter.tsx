import * as React from "react";
import {
    IconButton,
    IContextualMenuProps,
    IContextualMenuItem,
    IContextualMenuStyles,
    DirectionalHint,
} from "@fluentui/react";
import { NeutralColors } from "@fluentui/theme";

export interface ILayerCheckStates {
    text: boolean;
    tables: boolean;
    selectionMarks: boolean;
}

interface ILayerFilterProps {
    disabled: boolean;
    checkStates: ILayerCheckStates;
    modelType?: string;
    onItemClick: (item: IContextualMenuItem) => void;
}

export class LayerFilter extends React.PureComponent<ILayerFilterProps> {
    private filterStyles: Partial<IContextualMenuStyles> = {
        title: {
            background: "transparent",
            fontSize: 12,
            fontWeight: 600,
            padding: 10,
            borderBottom: `1px solid ${NeutralColors.gray30}`,
        },
    };

    private layerOptions = [
        {
            key: "text",
            text: "Text",
            iconProps: { iconName: "TextField" },
        },
        {
            key: "tables",
            text: "Tables",
            iconProps: { iconName: "Table" },
        },
        {
            key: "selectionMarks",
            text: "Selection marks",
            iconProps: { iconName: "CheckboxComposite" },
        },
    ];

    private getLayerOptions = () => {
        return this.layerOptions;
    };

    private handleItemClick = (ev, item?: IContextualMenuItem) => {
        ev?.preventDefault();
        if (item) {
            this.props.onItemClick(item);
        }
        return false;
    };

    private makeFilterMenu = (): IContextualMenuProps => {
        return {
            title: "Show extracted",
            ariaLabel: "Show extracted",
            styles: this.filterStyles,
            items: this.getLayerOptions().map((option) => ({
                ...option,
                canCheck: true,
                checked: this.props.checkStates[option.key] === true,
                iconProps: option.iconProps,
            })),
            onItemClick: this.handleItemClick,
            directionalHint: DirectionalHint.bottomRightEdge,
        };
    };

    public render() {
        return (
            <IconButton
                title="Choose extracted object to display"
                disabled={this.props.disabled}
                styles={{ rootDisabled: { backgroundColor: "transparent" } }}
                iconProps={{ iconName: "MapLayers" }}
                menuProps={this.makeFilterMenu()}
            />
        );
    }
}

export default LayerFilter;
