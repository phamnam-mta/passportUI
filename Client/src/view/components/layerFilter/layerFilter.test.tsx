import * as React from "react";
import { shallow } from "enzyme";
import { LayerFilter } from "./layerFilter";

describe("<LayerFilter />", () => {
    let baseProps;

    beforeEach(() => {
        baseProps = {
            checkStates: {
                text: true,
                tables: true,
                selectionMarks: true,
            },
            onItemClick: jest.fn(),
        };
    });

    describe("Rendering", () => {
        it("should render correctly", () => {
            const wrapper = shallow(<LayerFilter {...baseProps} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        it("should update filter state, when a filter option is clicked", () => {
            const mockItem = { key: "text", checked: true };
            const event = { preventDefault: jest.fn() };

            const wrapper = shallow(<LayerFilter {...baseProps} />);
            const button = wrapper.find("CustomizedIconButton") as any;
            button.prop("menuProps").onItemClick(event, mockItem);

            expect(event.preventDefault).toBeCalledTimes(1);
            expect(baseProps.onItemClick).toBeCalledTimes(1);
            expect(baseProps.onItemClick).toBeCalledWith(mockItem);
        });

        it("should not update filter state, when no filter option is clicked", () => {
            const event = { preventDefault: jest.fn() };

            const wrapper = shallow(<LayerFilter {...baseProps} />);
            const button = wrapper.find("CustomizedIconButton") as any;
            button.prop("menuProps").onItemClick(event, null);

            expect(event.preventDefault).toBeCalledTimes(1);
            expect(baseProps.onItemClick).not.toBeCalled();
        });
    });
});
