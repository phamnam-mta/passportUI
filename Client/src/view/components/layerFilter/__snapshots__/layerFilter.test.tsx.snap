// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LayerFilter /> Rendering should render correctly 1`] = `
<CustomizedIconButton
  iconProps={
    Object {
      "iconName": "MapLayers",
    }
  }
  menuProps={
    Object {
      "ariaLabel": "Show extracted",
      "directionalHint": 6,
      "items": Array [
        Object {
          "canCheck": true,
          "checked": true,
          "iconProps": Object {
            "iconName": "TextField",
          },
          "key": "text",
          "text": "Text",
        },
        Object {
          "canCheck": true,
          "checked": true,
          "iconProps": Object {
            "iconName": "Table",
          },
          "key": "tables",
          "text": "Tables",
        },
        Object {
          "canCheck": true,
          "checked": true,
          "iconProps": Object {
            "iconName": "CheckboxComposite",
          },
          "key": "selectionMarks",
          "text": "Selection marks",
        },
      ],
      "onItemClick": [Function],
      "styles": Object {
        "title": Object {
          "background": "transparent",
          "borderBottom": "1px solid #edebe9",
          "fontSize": 12,
          "fontWeight": 600,
          "padding": 10,
        },
      },
      "title": "Show extracted",
    }
  }
  styles={
    Object {
      "rootDisabled": Object {
        "backgroundColor": "transparent",
      },
    }
  }
  title="Choose extracted object to display"
/>
`;
